### PR TITLE
kicad-unstable: 2022-06-21 -> 2022-09-18

### LINKED ISSUE #191805: Kicad fails to build on unstable
### Steps To Reproduce
1. build kicad (tested on da6a05816e7fa5226c3f61e285ef8d9dfc868f3c)

### Build log

The relevant part of the build log is:

```
/build/source/include/gal/opengl/kiglew.h:53:14: error: #error "KICAD_USE_EGL must be defined since wxWidgets has been compiled with the EGL canvas"
   53 |             #error "KICAD_USE_EGL must be defined since wxWidgets has been compiled with the EGL canvas"
      |              ^~~~~
```

Complete build log is here:

https://gist.github.com/JCapucho/8ea9b13634acac8e96b92f2be20c80fd

### Additional context
Setting the `KICAD_USE_EGL` flag in the cmake build flags, solves the issue.

```
diff --git a/pkgs/applications/science/electronics/kicad/base.nix b/pkgs/applications/science/electronics/kicad/base.nix
index aa5d9a7ade1..c8457fa4514 100644
--- a/pkgs/applications/science/electronics/kicad/base.nix
+++ b/pkgs/applications/science/electronics/kicad/base.nix
@@ -78,6 +78,7 @@ stdenv.mkDerivation rec {
   cmakeFlags = [
     # RPATH of binary /nix/store/.../bin/... contains a forbidden reference to /build/
     "-DCMAKE_SKIP_BUILD_RPATH=ON"
+    "-DKICAD_USE_EGL=ON"
   ]
   ++ optionals (withScripting) [
     "-DKICAD_SCRIPTING_WXPYTHON=ON"

```

### Notify maintainers

@evils @Kiwi 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"

 - system: `"x86_64-linux"`
 - host os: `Linux 5.19.8, NixOS, 22.11 (Raccoon), 22.11pre-git`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - nixpkgs: `/nix/store/8s4q1xjnpkklhw3lgcxlvab9g6kln4qk-source`

```

### PR DESCRIPTION
**Pull request by evils**
###### Description of changes

fixes #191805 #192609
that fix may affect kicad (stable) as well, and since `EGL` has some association with wayland and i'm using xwayland on sway,
i'd like it to be reviewed on X11

also includes a fix to the update script, minor cleanup
and an update that required some more minor changes (most notably adding a dependency and changing another)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/electronics/kicad/base.nix [hunk 1]
```diff
@@ -151,15 +153,19 @@ stdenv.mkDerivation rec {
...
   ]
+  ++ optional (!stable) unixODBC
   ++ optional (withScripting) wxPython
   ++ optional (withNgspice) libngspice
   ++ optional (withOCC) opencascade-occt
   ++ optional (debug) valgrind
   ;
 
+  # started becoming necessary halfway into 2022, not sure what changed to break a test...
+  preInstallCheck = optionals (withNgspice) [ "export LD_LIBRARY_PATH=${libngspice}/lib" ];
```
Comment (wentasah) [inline pkgs/applications/science/electronics/kicad/base.nix:168]: This probably means that this library is somehow needed at build time. I'm not sure, but maybe adding the library also to `nativeBuildInputs` would have the same effect (not tested).

**Review by wentasah — COMMENTED** [review]
I've tested it on X11 and everything works. Thanks.

File: pkgs/applications/science/electronics/kicad/base.nix (hunk 1 above)
Comment (evils) [reply pkgs/applications/science/electronics/kicad/base.nix:168]: fair point, i didn't actually test the spice simulation part, as i'm not familiar with that...

Comment (evils) [reply pkgs/applications/science/electronics/kicad/base.nix:168]: adding the library to nativeBuildInputs instead setting it for the installCheck phase results in the same failure

going to give spice simulation a go next

Comment (evils) [reply pkgs/applications/science/electronics/kicad/base.nix:168]: simulation seems to work

i should probably open an issue about it not finding the library...

Comment (evils) [reply pkgs/applications/science/electronics/kicad/base.nix:168]: https://gitlab.com/kicad/code/kicad/-/issues/12491

**Review by stigtsp — COMMENTED** [review]
Build failed for `kicad-unstable-small`

Result of `nixpkgs-review pr 191887` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>kicad-unstable-small</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>kicad-small</li>
    <li>python310Packages.kicad</li>
    <li>python39Packages.kicad</li>
  </ul>
</details>


```
In file included from /build/source/qa/schematic_utils/schematic_file_util.cpp:32:
/build/source/eeschema/sch_plugins/kicad/sch_sexpr_parser.h:36:10: fatal error: schematic_lexer.h: No such file or directory
   36 | #include <schematic_lexer.h>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by evils** [comment]
> fatal error: schematic_lexer.h

i've seen this on a nixpkgs-review too

it didn't do it on a second attempt...

**Comment by wentasah** [comment]
Result of `nixpkgs-review pr 191887` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>kicad-small</li>
    <li>kicad-unstable-small</li>
    <li>python310Packages.kicad</li>
    <li>python39Packages.kicad</li>
  </ul>
</details>

**Comment by wentasah** [comment]
I confirm that the second nixpkgs-review works (see above). The missing header file `schematic_lexer.h` is generated by `cmake` during the build (see [eeschema/CMakeLists.txt](https://gitlab.com/kicad/code/kicad/blob/28f836ab5cd233e096ea49260fb3868874255508/eeschema/CMakeLists.txt#L601)). My guess is that there is some unspecified dependency within the `CMakeLists.txt` file, which makes the parallel build fail in the case, when `schematic_lexer.h` is created only after some other branch of the parallel build needs it. I think, we should report this upstream. I try to investigate the reason more in detail later.

For now, I think this PR can be merged. Hopefully, someone will restart the hydra job if it fails there.

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 191887` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>kicad-small</li>
    <li>kicad-unstable-small</li>
    <li>python310Packages.kicad</li>
    <li>python39Packages.kicad</li>
  </ul>
</details>

**Review by superherointj — APPROVED** [review]

**Comment by wentasah** [comment]
FYI: I think parallel build failures can be fixed by https://gitlab.com/kicad/code/kicad/-/merge_requests/1319.

**Pull request merged**